src/mlpack/core/optimizers/fw/atoms.hpp

```diff
@@ -0,0 +1,251 @@
+   * see the "Enhancement step" of the following paper:
+   * @code
+   * @article{RaoShaWri:2015Forward--backward,
+   *    Author = {Rao, Nikhil and Shah, Parikshit and Wright, Stephen},
+   *    Journal = {IEEE Transactions on Signal Processing},
+   *    Number = {21},
+   *    Pages = {5798--5811},
+   *    Publisher = {IEEE},
+   *    Title = {Forward--backward greedy algorithms for atomic norm regularization},
+   *    Volume = {63},
+   *    Year = {2015}
+   * }
+   * @endcode
+   *
+   * @param function function to be minimized.
+   * @param tau atom norm constraint.
+   * @param stepSize step size for projected gradient method.
+   * @param maxIteration maximum iteration number.
+   * @param tolerance tolerance for projected gradient method.
+   */
+  template<typename FunctionType>
+  void ProjectedGradientEnhancement(FunctionType& function,
+                                    double tau,
+                                    double stepSize,
+                                    size_t maxIteration = 100,
+                                    double tolerance = 1e-3)
+  {
+    arma::mat x;
+    RecoverVector(x);
+    double value = function.Evaluate(x);
+
+    for (size_t iter = 1; iter<maxIteration; iter++)
+    {
+      // Update currentCoeffs with gradient descent method.
+      arma::mat g;
+      function.Gradient(x, g);
+      g = currentAtoms.t() * g;
+      currentCoeffs = currentCoeffs - stepSize * g;
+
+      // Projection of currentCoeffs to satisfy the atom norm constraint.
+      ProjectionToL1(tau);
+
+      RecoverVector(x);
+      double valueNew = function.Evaluate(x);
+
+      if ((value - valueNew) < tolerance)
+        break;
+
+      value = valueNew;
+    }
+  }
+
+
+  //! Get the current atom coefficients.
+  const arma::vec& CurrentCoeffs() const { return currentCoeffs; }
+  //! Modify the current atom coefficients.
+  arma::vec& CurrentCoeffs() { return currentCoeffs; }
+
+  //! Get the current atoms.
+  const arma::mat& CurrentAtoms() const { return currentAtoms; }
+  //! Modify the current atoms.
+  arma::mat& CurrentAtoms() { return currentAtoms; }
+
+ private:
+  //! Coefficients of current atoms.
+  arma::vec currentCoeffs;
+
+  //! Current atoms in the solution space.
+  arma::mat currentAtoms;
+
+
+  /**
+   * Projection of currentCoeffs to L1 ball with norm tau.
+   * Used in ProjectedGradientEnhancement().
+   *
+   * See the paper:
+   * @code
+   * @inproceedings{DucShaSin:2008Efficient,
+   *    Author = {Duchi, John and Shalev-Shwartz, Shai and Singer, Yoram and Chandra, Tushar},
+   *    Booktitle = {Proceedings of the 25th international conference on Machine learning},
+   *    Organization = {ACM},
+   *    Pages = {272--279},
+   *    Title = {Efficient projections onto the l 1-ball for learning in high dimensions},
+   *    Year = {2008}}
+   * @endcode
+   *
+   * @param tau atom norm constraint.
+   */
+  void ProjectionToL1(const double tau)
```

Comment:
Furthermore, once we move it to somewhere shared, we can write a unit test for it.